### ISSUE TITLE
Suggestions, maybe a bug in the docs 

### ISSUE THREAD

**Issue by ankopainting** [ConvertKit/excess_flow#1]
> ExcessFlow is a high precision Redis based rate limiter; it means that even with hundreds and even thousands requests coming in all at once it will not allow an occasional request slip over limit (causing potential race conditions or unwanted extra invocations of your code).

I just implemented a sliding window rate limiter myself, with redis in ruby so I just wanted to give some thoughts and ideas.  Overall i'd say I really like this gem and the layout of the code is well done.  My code is not publicly available for reasons I won't get into.

I'm also being pretty nitpicky, so sorry for that.

I think the statement "it means that even with hundreds and even thousands requests coming in all at once it will not allow an occasional request slip over limit" is false, because the Timestamps in a given ordered set are determined by the client, and not guaranteed to be unique.  

You do
```
r.zadd(
          configuration.counter_key,
          current_timestamp,
          current_timestamp
        )
```

From the redis docs;
If a specified member is already a member of the sorted set, the score is updated and the element reinserted at the right position to ensure the correct ordering.

So if 2 clients have some clock drift (most clocks are slightly out), and when they each acquire the global mutex they both set the same value for current_timestamp, then the nature of the set is that it will update the same element to the same score, in effect not adding the current request to your set.

The fix I used in my code was setting the set value to a guid, similar to;
```
r.zadd(
          configuration.counter_key,
          current_timestamp,
          SecureRandom.uuid
        )
```
so it's always going to be unique and the rest of your operations work.  timestamps set in the future (relative to your first instance) will still be counted by your zcount, so there is a slight chance you'll block an extra request occasionally, but the chance is small.

The second thing i just wanted to comment on (sorry for the nitpick), is the mutex is unnecessary.  Instead you could use a redis transaction. Here is what I'm doing;

```
    ret = redis.pipelined do
      redis.zremrangebyscore full_key, 0, clear_before_timestamp # remove requests past period
      redis.zcard full_key # count of how many requests we've received
      redis.zadd full_key, current_timestamp, SecureRandom.uuid
      redis.expire full_key, period
    end
```

The redis docs say;
All the commands in a transaction are serialized and executed sequentially. It can never happen that a request issued by another client is served in the middle of the execution of a Redis transaction. This guarantees that the commands are executed as a single isolated operation.

https://redis.io/topics/transactions

Just wanted to point out another way to do it :)

Thank you for your library and your time.  I might have got something wrong so please correct me if i have.

**Comment by ryakh** [comment]
@ankopainting thanks for bringing those up! 

I've fixed the issue with possible duplicate requests being caused by clock drift in this commit:
https://github.com/ConvertKit/excess_flow/commit/8c63fb5b3a7cd44ae28d072269bc9f1c03ddbfba

This should be fixed in `1.0.2` that I've just released.

As per your suggestion to use Redis' transaction instead of `GlobalMutex`; that's something I need to explore and test a bit more. Namely I want to see how it will fit into other rate-limiting strategies that we might be adding in the future. If it will work out it will be great as we will be using more native API's to solve this problem instead of keeping a code around that requires maintenance.

Would you feel comfortable if I go ahead and close this issue and open a new one that will concern only exploring using transactions instead of `GlobalMutex`?

**Comment by ankopainting** [comment]
Of course, it's your library, just trying to help :)
